src/position/decrease_position_collateral_utils.cairo

```diff
@@ -43,86 +42,491 @@ struct PayForCostResult {
+    // price_impact_diff_usd is the difference between the maximum price impact and the originally calculated price impact
+    // e.g. if the originally calculated price impact is -$100, but the capped price impact is -$80
+    // then priceImpactDiffUsd would be $20
+    let (price_impact_usd_, price_impact_diff_usd_, execution_price_) = get_execution_price(
+        params, cache.prices.index_token_price
+    );
+    values.price_impact_usd = price_impact_usd_;
+    values.price_impact_diff_usd = price_impact_diff_usd_;
+    values.execution_price = execution_price_;
+    // the total_position_pnl is calculated based on the current indexTokenPrice instead of the executionPrice
+    // since the executionPrice factors in price impact which should be accounted for separately
+    // the sizeDeltaInTokens is calculated as position.size_in_tokens() * size_delta_usd / position.size_in_usd()
+    // the basePnlUsd is the pnl to be realized, and is calculated as:
+    // total_position_pnl * size_delta_in_tokens / position.size_in_tokens()
+    let (base_pnl_usd_, uncapped_base_pnl_usd_, size_delta_in_tokens_) =
+        position_utils::get_position_pnl_usd(
+        params.contracts.data_store,
+        params.market,
+        cache.prices,
+        params.position,
+        params.order.size_delta_usd
+    );
+    values.base_pnl_usd = base_pnl_usd_;
+    values.uncapped_base_pnl_usd = uncapped_base_pnl_usd_;
+    values.size_delta_in_tokens = size_delta_in_tokens_;
+
+    let get_position_fees_params: position_pricing_utils::GetPositionFeesParams =
+        position_pricing_utils::GetPositionFeesParams {
+        data_store: params.contracts.data_store,
+        referral_storage: params.contracts.referral_storage,
+        position: params.position,
+        collateral_token_price: cache.collateral_token_price,
+        for_positive_impact: values.price_impact_usd > 0,
+        long_token: params.market.long_token,
+        short_token: params.market.short_token,
+        size_delta_usd: params.order.size_delta_usd,
+        ui_fee_receiver: params.order.ui_fee_receiver,
     };
-    let price = Price { min: 0, max: 0, };
-    let position_fees = PositionFees {
-        referral: position_referral_fees,
-        funding: position_funding_fees,
-        borrowing: position_borrowing_fees,
-        ui: position_ui_fees,
-        collateral_token_price: price,
-        position_fee_factor: 0,
-        protocol_fee_amount: 0,
-        position_fee_receiver_factor: 0,
-        fee_receiver_amount: 0,
-        fee_amount_for_pool: 0,
-        position_fee_amount_for_pool: 0,
-        position_fee_amount: 0,
-        total_cost_amount_excluding_funding: 0,
-        total_cost_amount: 0,
+
+    let mut fees: position_pricing_utils::PositionFees = position_pricing_utils::get_position_fees(
+        get_position_fees_params
+    );
+
+    // if the pnl is positive, deduct the pnl amount from the pool
+    if values.base_pnl_usd > 0 {
```

Comment:
add this  `// use pnlTokenPrice.max to minimize the tokens paid out` after the if statement